### DISCUSSION TITLE
Why do we set the range to be ignored one-at-a-time?

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#78] [Q&A] [answered]
In this bit:
https://github.com/congo-cc/congo-parser-generator/blob/6fa4f254911d7158d3b2c5e0922aea1eaac56637/src/templates/java/TokenSource.java.ftl#L263-L266
why is there a `for` loop rather than just a call to `setIgnoredRange(offset, nlIndex)`?

**Comment by revusky** [comment, marked as answer]
Good question. I think the answer is likely that I was just sleep deprived or something. Or maybe I was working from a patch given to me by the guy who wanted this feature (Gravelbones) and he wrote that originally. It hardly matters.

Though, actualy, I think the loop would need to be replaced by `setIgnoredRange(offset, nlIndex+1)` since that method is defined with the endpoint non-inclusive. The original conversation that resulted in this feature is here: https://github.com/javacc21/javacc21/pull/163 and actually, I would have to assume that Gravelbones is the only person using this.  Or nobody is...

**Reply by vsajip** [reply]
I guess it could come in handy for C/C++ parsing ...

**Comment by revusky** [comment]
> I guess it could come in handy for C/C++ parsing ...

Well, yeah. That's why the guy requested the feature. I don't think it's documented anywhere, so you'd actually have to find it by reading the code. Or finding the discussion there, actually... But, you know, generally, there is a lot of machinery there that for general parsing problems, like we just happen to have this C_CONTINUATION_LINE setting because a guy asked for it. That, in turn, was trivial to implement because you can just use that sort of preprocessor functionality, turning on/off sections of the input, the functionality is more or less exposed, assuming  you set `USES_PREPROCESSOR=true` and then you have that `setIgnoredRange` and `setUnparsedLines` sort of machinery available.

You know, actually, as opposed to cleaning up these glitches in the code, which has some value, there is more value in documenting all these things. Though, admittedly, that is not a separate thing exactly, since the main way to know about this stuff is to read over the code, but then, yeah, you see these glitches and it's natural to want to fix them,... But you know, telling people that these things are available... Well, like I say, you can figure out they're available by reading the code, I guess, like looking at examples/preprocessor, but even then, you have to know where to look, so....
